tensorpipe/channel/cuda/channel.cc

```diff
@@ -0,0 +1,406 @@
+                       << ")";
+            TP_DCHECK_CUDA(cudaEventDestroy(startEv));
+            TP_DCHECK_CUDA(cudaStreamWaitEvent(impl.stream_, stopEv, 0));
+            TP_DCHECK_CUDA(cudaEventDestroy(stopEv));
+            callback(impl.error_);
+          }));
+}
+
+// Receive memory region from peer.
+void Channel::recv(
+    TDescriptor descriptor,
+    void* ptr,
+    size_t length,
+    TRecvCallback callback) {
+  impl_->recv(std::move(descriptor), ptr, length, std::move(callback));
+}
+
+void Channel::Impl::recv(
+    TDescriptor descriptor,
+    void* ptr,
+    size_t length,
+    TRecvCallback callback) {
+  loop_.deferToLoop([this,
+                     descriptor{std::move(descriptor)},
+                     ptr,
+                     length,
+                     callback{std::move(callback)}]() mutable {
+    recvFromLoop_(std::move(descriptor), ptr, length, std::move(callback));
+  });
+}
+
+void Channel::Impl::recvFromLoop_(
+    TDescriptor descriptor,
+    void* ptr,
+    size_t length,
+    TRecvCallback callback) {
+  TP_DCHECK(loop_.inLoop());
+
+  const uint64_t sequenceNumber = nextTensorBeingReceived_++;
+  TP_VLOG(4) << "Channel " << id_ << " received a recv request (#"
+             << sequenceNumber << ")";
+  callback = [this, sequenceNumber, callback{std::move(callback)}](
+                 const Error& error) {
+    TP_VLOG(4) << "Channel " << id_ << " is calling a recv callback (#"
+               << sequenceNumber << ")";
+    callback(error);
+    TP_VLOG(4) << "Channel " << id_ << " done calling a recv callback (#"
+               << sequenceNumber << ")";
+  };
+
+  if (error_) {
+    callback(error_);
+    return;
+  }
+
+  proto::Descriptor pbDescriptor;
+  loadDescriptor(pbDescriptor, descriptor);
+
+  const cudaIpcMemHandle_t* remoteHandle =
+      reinterpret_cast<const cudaIpcMemHandle_t*>(
+          pbDescriptor.handle().c_str());
+
+  const cudaIpcEventHandle_t* startEvHandle =
+      reinterpret_cast<const cudaIpcEventHandle_t*>(
+          pbDescriptor.start_ev_handle().c_str());
+  const cudaIpcEventHandle_t* stopEvHandle =
+      reinterpret_cast<const cudaIpcEventHandle_t*>(
+          pbDescriptor.stop_ev_handle().c_str());
+
+  cudaEvent_t startEv;
+  TP_DCHECK_CUDA(cudaIpcOpenEventHandle(&startEv, *startEvHandle));
+  cudaEvent_t stopEv;
+  TP_DCHECK_CUDA(cudaIpcOpenEventHandle(&stopEv, *stopEvHandle));
+
+  TP_VLOG(6) << "Channel " << id_ << " is copying payload (#" << sequenceNumber
+             << ")";
+
+  TP_DCHECK_CUDA(cudaStreamWaitEvent(stream_, startEv, 0));
+
+  // Perform copy.
+  if (length > 0) {
+    void* remotePtr;
+    TP_DCHECK_CUDA(cudaIpcOpenMemHandle(
+        &remotePtr, *remoteHandle, cudaIpcMemLazyEnablePeerAccess));
+
+    TP_DCHECK_CUDA(cudaMemcpyAsync(ptr, remotePtr, length, cudaMemcpyDefault));
+    TP_DCHECK_CUDA(cudaIpcCloseMemHandle(remotePtr));
+  }
+  TP_DCHECK_CUDA(cudaEventRecord(stopEv));
```

Comment:
I think the receiver here must also free the events with `cudaEventDestroy(startEv)`.

Reply:
I don't think so. The docs states:
> Performing operations on the imported event after the exported event has been freed with cudaEventDestroy will result in undefined behavior.

The event itself "lives" on the device, and the `cudaEvent_t` variables are just handles.

Reply:
If you look at the doc for `cudaIpcOpenEventHandle` you'll see it says "This event must be freed with cudaEventDestroy".